File: backend/src/main/java/com/example/Flicktionary/domain/movie/entity/Movie.java
package com.example.Flicktionary.domain.movie.entity;

import com.example.Flicktionary.domain.director.entity.Director;
import com.example.Flicktionary.domain.genre.entity.Genre;
import com.example.Flicktionary.domain.review.entity.Review;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

@Entity
@Getter
@Setter
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class Movie {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "tmdb_id", nullable = false, unique = true)
    private Long tmdbId;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String overview;

    private LocalDate releaseDate;

    private LocalDate fetchDate;

    private String status;

    private String posterPath;

    private Integer runtime;

    private String productionCountry;

    private String productionCompany;

    private double averageRating;

    private int ratingCount;

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "movie_genre",
            joinColumns = @JoinColumn(name = "movie_id"),
            inverseJoinColumns = @JoinColumn(name = "genre_id")
    )
    @Builder.Default
    private List<Genre> genres = new ArrayList<>();

    @OneToMany(mappedBy = "movie", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<MovieCast> casts = new ArrayList<>();

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "director_id")
    private Director director;

    @OneToMany(mappedBy = "movie", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<Review> reviews = new ArrayList<>();
}

File: backend/src/main/java/com/example/Flicktionary/domain/movie/entity/Movie.kt
package com.example.Flicktionary.domain.movie.entity

import com.example.Flicktionary.domain.director.entity.Director
import com.example.Flicktionary.domain.genre.entity.Genre
import jakarta.persistence.*
import java.time.LocalDate

@Entity
class Movie(
    @Column(nullable = false)
    var title: String,
    @Column(columnDefinition = "TEXT")
    var overview: String,
    var releaseDate: LocalDate?,
    var status: String,
    var posterPath: String?,
    var runtime: Int,
    var productionCountry: String,
    var productionCompany: String
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private var _id: Long? = null

    var id: Long
        get() = _id ?: 0
        set(value) {
            _id = value
        }

    var averageRating = 0.0

    var ratingCount = 0

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
        name = "movie_genre",
        joinColumns = [JoinColumn(name = "movie_id")],
        inverseJoinColumns = [JoinColumn(name = "genre_id")]
    )
    val genres: MutableList<Genre> = mutableListOf()

    @OneToMany(mappedBy = "movie", cascade = [CascadeType.ALL], orphanRemoval = true)
    val casts: MutableList<MovieCast> = mutableListOf()

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "director_id")
    var director: Director? = null
}
